File: 1161CharlyandNito/charlyandnito.py

try:
    import Queue as queue
except ImportError:
    import queue 


class Vertice: 
    def __init__(self ,clave):
        self.id = clave 
        self.conectadoA = {}

    def agregarVecino (self, vecino , ponderacion = 0):
        self.conectadoA [vecino] = ponderacion 

    def __str__(self):
        return str(self.id) + ' conectadoA: ' + str([x.id for x in self.conectadoA])

    def obtenerConexiones (self):
        return self.conectadoA.keys()

    def obtenerId (self):
        return self.id 

    def obtenerPonderacion (self , vecino ):
        return self.conectadoA[vecino]


class  Grafo:
    def __init__(self):
        self.listadeVertices = {} 
        self.numVertices = 0 

    def agregarVertice (self, clave):
        self.numVertices = self.numVertices + 1 
        nuevoVertice = Vertice (clave) 
        self.listadeVertices[clave] = nuevoVertice
        return nuevoVertice

    def obtenerVertice (self,n):
        if n in self.listadeVertices:
            return self.listadeVertices[n]
        else:
            return None

    def __contains__ (self , n):
        return n in self.listadeVertices

    def agregarArista (self , de , a , costo = 0):
        if de not  in self.listadeVertices:
            nv = self.agregarVertice (de) 
        if a not in self.listadeVertices:
            nv = self.agregarVertice (a) 
        self.listadeVertices[de].agregarVecino(self.listadeVertices[a] , costo) 

    def obtenerVertices (self):
        return self.listadeVertices.keys() 

    def __iter__(self):
        return iter (self.listadeVertices.values())  


priority_queue = queue.PriorityQueue()


MAX = 5001
INF = 1<<30 


g = Grafo() 
distanciaC = [] 
distanciaN = [] 
distanciaB = [] 
visitado = [] 
J,B,C,N,S = 0,0,0,0,0

def init  (distancia ):
    visitado.clear()
    i = 0 
    while i < MAX : 
        distancia.append (INF) 
        visitado.append (False) 
        i = i + 1 

def relajacion (actual , adyacente , peso , distancia):
    if distancia[actual] + peso < distancia[adyacente]:
        distancia[adyacente] = distancia[actual] + peso 
        priority_queue.put((distancia[adyacente],adyacente))


def dijkstra (inicial ,distancia):
    init(distancia)
    distancia[inicial] = 0
    priority_queue.put((0,inicial))
    
    while not priority_queue.empty():
        top = priority_queue.get()
        actual = top[1]
        vertice_actual = g.obtenerVertice(actual)
        if visitado[actual]: continue 
        visitado[actual] = True

        for adyacente in vertice_actual.obtenerConexiones():
            peso = vertice_actual.obtenerPonderacion(adyacente)

            if not visitado[adyacente.obtenerId()]:
                relajacion (actual , adyacente.obtenerId() , peso , distancia)       


def solucion ():
    distC = distanciaB[C-1] 
    distN = distanciaB[N-1]
    j = 0 
    c = distC
    n = distN

    for i in range (J) :
        if distC == distanciaB[i] + distanciaC[i] and distN == distanciaB[i] + distanciaN[i] and distanciaB[i] > j :
            j = distanciaB[i]
            c = distanciaC[i]
            n = distanciaN[i] 

    print (str (j) + ' ' + str (c) + ' ' +str(n))


if __name__ == '__main__':
    
    while True :
        line  = str(input()).split()
        
        J= int(line[0])
        B= int(line[1])
        C= int(line[2])
        N= int(line[3])
        S= int(line[4])

        if J == -1 and B == -1 and C ==-1 and N ==-1 and S ==-1:
            break 


        for v in range (J):
            g.agregarVertice(v)

        i=0
        while(i< S):
            line = str(input()).split()
            E1 ,E2 ,L = int(line[0]) , int (line [1]) , int (line[2])
            g.agregarArista(E1-1,E2-1,L)
            g.agregarArista(E2-1,E1-1,L)
            i = i + 1
        
        #for v in g : 
        #    for w in v.obtenerConexiones():
        #        print ( "( %s , %s , %s )" % (v.obtenerId()+1 , w.obtenerId()+1 , v.obtenerPonderacion(w)))
        
        dijkstra(B-1,distanciaB)
        dijkstra(C-1,distanciaC)
        dijkstra(N-1,distanciaN)

        solucion() 

File: 1161CharlyandNito/charlyandnito.cpp
#include <bits/stdc++.h> 

using namespace std ; 

#define MAX 5001 /* aqui defino la mayor cantidad de nodos que puede tener mi grafo */ 
#define Node pair <int, int> /* esto es para usarlo en la cola de prioridad para el algoritmo de dijksta*/ 
#define INF 1 << 30 /* aqui defino el infinito para obtener una cota superior de las distancias que pueden tener las aristas del grafo */ 

int J,B,C,N,S ; /*para leer los datos de la entrada estandar */
vector <Node> ady [MAX] ;/* este es el grafo como una lista de adyacencia */
int distanciaC [MAX]; /*la distancia de la casa de CHARLY a las demas nodos del grafo */
int distanciaN [MAX]; /*la distancia de la casa de NITO a las demas nodos del grafo */
int distanciaB [MAX]; /*la distancia de del BAR a las demas nodos del grafo */
set< Node > Q ;   /*una cola de prioridad que usa el comarador de arriba para procesar los nodos en el algoritmo de dijkstra */

void init (int distancia []) /* esto es para inicializar los arrays que uso en el algoritmo de dijkstra */
{
    for (int i = 0; i < J; i++)
    {
        distancia[i] = INF;  /* pongo todas las distancias a infinito (un valor mayor del que toma la arista de mayor peso)*/
    }
}

void shortedPath (int start , int distance []) 
{
    init(distance);  /*aqui inicializo el array donde voy a poner las distancias */
    Q.insert(make_pair(0,start));  /*inseto el primer elemanto en la cola de prioridad con distancia 0 */
    distance[start] = 0;   /*pongo la distancia en el nodo inicial en 0 */

    while (! Q.empty())
    {
        Node tmp = * (Q.begin());  /*en tmp pongo el primer elemento del 
                                    principio de la cola (el que menor peso tiene )*/
        Q.erase(Q.begin());  /*borro el primer elemnto del conjunto*/

        int curr = tmp.second ;   /*en curr pongo el numero del nodo que voy a procesar*/
        vector <Node> :: iterator i ; /*declaro un iterador para recorrer los adyacentes */
        for(i = ady[curr].begin() ; i != ady[curr].end() ; ++i)
        {
            int v = (*i).first ;    /*en v pongo el id del vecino que voy a procesar*/
            int weight = (*i).second;  /* pongo en weight el peso del vecino*/

            if (distance[curr] + weight < distance[v])  /* si la distancia hasta  el adyacente es 
                                                        mayor que la distancia del actual mas es peso
                                                        de la arista que estoy procesando entonces actualizo  */
            {
                if (distance[v] < INF) /*si  el adyacente ya estaba en la cola entonces lo saco de la cola  */
                {
                    Q.erase(Q.find(make_pair(distance[v] , v))); 
                } 
                distance[v] = distance[curr] + weight ; /*actualizo la distancia */  
                Q.insert(make_pair(distance[v],v)); /* pongo en vertice con su nuevo peso en la cola
                                                     para ser procesado */
            } 
        }
    }
}


void solucion()
{   
    //auto start = chrono::system_clock::now(); 
    //unsigned t0 , t1; 
    //t0 = clock();


    shortedPath(B-1,distanciaB); /* aplicar dijkstra desde el BAR */ 
    shortedPath(C-1,distanciaC); /* aplicar dijkstra desde la casa de CHARLY*/
    shortedPath(N-1,distanciaN); /* aplicar dijkstra desde la casa de NITO */ 

    int distC = distanciaB[C-1]; /* esta es la menor distancia desde el BAR hasta la casa de CHARLY*/
    int distN = distanciaB[N-1]; /* esta es la menor distancia desde el BAR hasta la casa de NITO*/
    int j = 0 , c =distC , n =distN; /* j es la mayor distancia que ellos caminan juntos , c es la distancia que camina CHARLY solo y n es la distancia que camina  NITO solo */

    for (int i = 0; i < J; i++)
    {
        /* aqui por cada vertice pregunto si este es un vertice comun a algun par de caminos minimos que van del BAR a casa de NITO y a casa de CHARLY 
           respectivamente y que la distancia desde ese nodo al BAR es mayor que una anterior analizada que cumpla las mismas condiciones de estar 
           en un camino de distancia minima*/
        if (distC == distanciaB[i]+distanciaC[i] && distN == distanciaB[i] + distanciaN[i] && distanciaB[i] > j)
        {
            j = distanciaB[i];  /* actualizo la mayor distancia que caminan juntos */
            c = distanciaC[i];  /* actualizo la distancia que camina CHARLY solo */
            n = distanciaN[i];  /* actualiza la distancia que camina NITO solo */ 
        }
    }
    //t1 = clock () ;
    //auto end = std::chrono::system_clock::now();
    //std::chrono::duration<double> elapsed = end - start;
    //std::cout << "Elapsed time: " << elapsed.count() << "s" <<endl;
    //double time = (double (t1-t0)/CLOCKS_PER_SEC); 
    cout << j << " " << c << " " << n << endl ; 
}

int main ()
{
     
    while (true)
    {
        cin >> J >> B >> C >> N >> S ; 
        if ( J == -1 && B ==-1 && C ==-1 && N ==-1 && S==-1)break; /* si todas la entradas fueron -1 entonces salgo del ciclo */

        int E1 , E2 , L; /* variables temporales par obtener las aristas y sus respectivos pesos */

        while (S-->0) /*leo todas las entradas  que me da la informacion de las aristas */
        {
            cin >> E1 >> E2 >> L; 
            ady[E1-1].push_back(Node(E2-1,L));
            ady[E2-1].push_back(Node(E1-1,L));
        }       

        solucion(); /* calcular la solucion al problema e imprimir la respuesta */ 
    }    


    return 0 ;     
}